sweepbatcher: add greedy batch selection algorithm

starius opened the pull request:
New algorithm is tried to select a batch to add a sweep to. It selects a batch for the sweep using the greedy algorithm, which minimizes costs, and adds the sweep to the batch or creates new batch, if it is the most cost efficient alternative. If the batch does not accept the sweep (too long timeout distance), next batch is tried. If the whole new approach fails, the old algorithm is used, trying to add to any batch, or starting a new batch.

The PR also includes some refactorings and fixes small inaccuracies in fee estimations.

#### Pull Request Checklist
- [ ] Update `release_notes.md` if your PR contains major features, breaking changes or bugfixes

lightninglabs-deploy commented:
@bhandras: review reminder
@sputn1ck: review reminder
@hieblmi: review reminder

bhandras commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,290 @@
+const specialId = -1
+
+// entity is either batch of sweep and it holds data important for selection.
+type entity struct {
```
I think `entity` is not necessarily the best name as it sounds a bit too general. Maybe consider something like `sweepSet` or `sweepEntity`? Also given that the role of this type is to store weight and fee information perhaps we can be even more specific: `sweepFeeDetails`?

starius replied:
I renamed `entity` to `feeDetails`.

bhandras commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,290 @@
+
+// greedyAddSweep selects a batch for the sweep using the greedy algorithm,
+// which minimizes costs, and adds the sweep to the batch.
+func (b *Batcher) greedyAddSweep(ctx context.Context, sweep *sweep) error {
```
I think some more details about the algorithm in the godoc can be be helpful.

starius replied:
Done

bhandras commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,290 @@
+}
+
+// estimateSweepWeight returns sweep's input weight and one sweep batch weight.
+func estimateSweepWeight(s *sweep) (*entity, *entity, error) {
```
nit: i think returning values instead of pointers is cleaner.

starius replied:
Done

bhandras commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,290 @@
+}
+
+// estimateSweepWeight returns sweep's input weight and one sweep batch weight.
+func estimateSweepWeight(s *sweep) (*entity, *entity, error) {
```
nit: perhaps naming this functions as `estimateSweepFeeDelta` or `estimateSweepFeeIncrement` could be more readable. Maybe the godoc can also be made clearer slightly.

starius replied:
Renamed to `estimateSweepFeeIncrement` and added more godoc:

```go
// estimateSweepFeeIncrement returns fee details for adding the sweep to a batch
// and for creating new batch with this sweep only.
func estimateSweepFeeIncrement(s *sweep) (feeDetails, feeDetails, error) {
```

bhandras reviewed (commented):
Looks good, just a few comments!

starius commented on sweep/sweeper.go:
```diff
@@ -207,6 +207,28 @@ func (s *Sweeper) GetSweepFeeDetails(ctx context.Context,
+
+	// Add output.
+	if err := AddOutputEstimate(&weightEstimate, destAddr); err != nil {
+		return 0, 0, 0, fmt.Errorf("estimate fee: %w", err)
```
Fixed. Also added error message in case adding input fails.

starius commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -848,6 +846,19 @@ func (b *Batcher) loadSweep(ctx context.Context, swapHash lntypes.Hash,
 			swapHash[:6], err)
 	}
 
+	var minFeeRate chainfee.SatPerKWeight
```
minFeeRate is 0 be default. If customFeeRate is not provided, then rbfCache.FeeRate is also 0 and method batch.updateRbfRate() updates it to current fee rate according to batchConfTarget.

Added a comment to that code about this.

bhandras replied:
SGTM

starius commented:
@bhandras Thanks for the review! 💯

I fixed the comments. Please take another look!

bhandras commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -848,6 +846,22 @@ func (b *Batcher) loadSweep(ctx context.Context, swapHash lntypes.Hash,
 			swapHash[:6], err)
 	}
 
+	// minFeeRate is 0 be default. If customFeeRate is not provided, then
```
nit: s/be/by

starius replied:
Thanks! Fixed

bhandras commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,306 @@
+	}
+
+	// Estimate new batch.
+	newBatch, err := estimateBatchWeight(batch)
```
nit: to avoid confusing names and improve readability, consider naming `newBatch` => `feeDetails1` and `batch2` => `feeDetails2` or something similar as these are not batches.

starius replied:
I renamed them to `fd1` and `fd2`.

bhandras commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,306 @@
+// details has also IsExternalAddr flag. There is a rule that sweeps having flag
+// IsExternalAddr must go in individual batches. Cooperative spending is only
+// available if all the sweeps support cooperative spending path.
+func selectBatch(batches []feeDetails, sweep, oneSweepBatch feeDetails) (int32, error) {
```
nit: line too long

starius replied:
Fixed

bhandras reviewed (approved):
LGTM 🎉

starius commented:
Updated commit "sweepbatcher: fix inaccuracies in fee estimations": function `publishBatch` also used `AddP2TROutput` regardless of destination address type. Fixed it, similarly to `publishBatchCoop` function: use function AddOutputEstimate to estimate output size based on address type.

starius commented:
Rebased

hieblmi commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -375,18 +375,22 @@ func (b *Batcher) Run(ctx context.Context) error {
 		case sweepReq := <-b.sweepReqs:
 			sweep, err := b.fetchSweep(runCtx, sweepReq)
 			if err != nil {
+				log.Warnf("fetchSweep failed: %v.", err)
```
👍

hieblmi commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,308 @@
+
+	// Add inputs.
+	for _, sweep := range batch.sweeps {
+		// TODO: it should be txscript.SigHashDefault.
```
why not sign the whole tx? Maybe leave a comment here for the rationale for `SigHashDefault`.

starius replied:
> why not sign the whole tx?

There is no reason to use non-standard sighash here. It was just done this way (SigHashAll) in fee estimation in `publishBatchCoop`, so I did it the same way in the greedy algorithm and left this TODO.

I replaced it later in this PR (in commit "sweepbatcher: fix inaccuracies in fee estimations") and removed the TODO.

hieblmi reviewed (commented):
Great effort to save fees for the users of loop out💰. I've left a few comments.
I will test this locally next.

hieblmi commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -465,6 +470,19 @@ func (b *Batcher) handleSweep(ctx context.Context, sweep *sweep,
 
+	// If custom fee rate provider is used, run the greedy algorithm of
+	// batch selection to minimize costs.
+	if b.customFeeRate != nil {
```
Also, how are you planning to enable the custom fee rates? Will this be a user configured parameter which will be passed in as option to `batcher := sweepbatcher.NewBatcher(`?

starius replied:
Great idea! We can always use it. I pushed two commits:

 - sweepbatcher: always fill sweep.minFeeRate
 - sweepbatcher: always try greedy batch selection

to implement this.

starius replied:
Updated commit "sweepbatcher: always fill sweep.minFeeRate". Previous version of it caused an extra fee bump and failure of test TestCustomSweepConfTarget. Fixed.

Also pushed commit "loopout_test: enable logging in sweepbatcher", it helped debugging the issue.

starius commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,308 @@
+	for _, batch := range batches {
+		// If either the batch or the sweep has IsExternalAddr flag,
+		// the sweep can't be added to the batch, so skip.
+		if batch.IsExternalAddr || sweep.IsExternalAddr {
```
In case of `batch.IsExternalAddr` we can't return `specialId` yet. The sweep can be without `IsExternalAddr` and there might be a batch without IsExternalAddr, which it can be added to.

I moved the check against `sweep.IsExternalAddr` out of the cycle, to the beginning of the function.

Renamed `specialId` to `newBatchSignal`.

hieblmi replied:
Sounds good!

starius commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,308 @@
+// selectBatch returns index of the batch adding to which minimizes costs.
+// For each batch its fee rate and two weight are provided: weight in case of
+// cooperative spending and weight in case non-cooperative spending (using
+// preimages instead of taproot key spend). Also a hint is provided to signal
```
Fixed

starius commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,308 @@
+}
+
+// selectBatch returns index of the batch adding to which minimizes costs.
+// For each batch its fee rate and two weight are provided: weight in case of
```
Fixed

starius commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,308 @@
+	}
+}
+
+// selectBatch returns index of the batch adding to which minimizes costs.
```
Fixed

starius commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,308 @@
+
+	// Collect weight and fee rate info about existing batches.
+	batches := make([]feeDetails, 0, len(b.batches))
+	for _, batch := range b.batches {
```
Renamed to `existingBatch`. Below in this method renamed another instance to `bestBatch`.

starius commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,308 @@
+	// Collect weight and fee rate info about existing batches.
+	batches := make([]feeDetails, 0, len(b.batches))
+	for _, batch := range b.batches {
+		newBatchFeeDetails, err := estimateBatchWeight(batch)
```
Renamed to `batchFeeDetails`.

starius commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,308 @@
+// it passes the data to selectBatch() function, which emulates adding the sweep
+// to each batch and creating new batch for the sweep, and calculates the costs
+// of each alternative. Based on the estimates of selectBatch(), this method
+// adds the sweep to a batch, or creates new batch for it. If this method fails
```
Fixed

starius commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -481,6 +499,12 @@ func (b *Batcher) handleSweep(ctx context.Context, sweep *sweep,
+// launchNewBatch creates new batch, starts it and adds the sweep to it.
+func (b *Batcher) launchNewBatch(ctx context.Context, sweep *sweep) error {
+	// Spin up a fresh batch.
 	batch, err := b.spinUpBatch(ctx)
```
Done

starius commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -481,6 +499,12 @@ func (b *Batcher) handleSweep(ctx context.Context, sweep *sweep,
+}
+
+// launchNewBatch creates new batch, starts it and adds the sweep to it.
+func (b *Batcher) launchNewBatch(ctx context.Context, sweep *sweep) error {
```
Done

starius commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -465,6 +470,19 @@ func (b *Batcher) handleSweep(ctx context.Context, sweep *sweep,
+	// If custom fee rate provider is used, run the greedy algorithm of
+	// batch selection to minimize costs.
+	if b.customFeeRate != nil {
+		err := b.greedyAddSweep(ctx, sweep)
```
Fixed. Now this code is not inside the `if` (see commit "sweepbatcher: always try greedy batch selection").

starius commented on sweepbatcher/sweep_batcher.go:
```diff
@@ -121,6 +121,11 @@ type SweepInfo struct {
 	// DestAddr is the destination address of the sweep.
 	DestAddr btcutil.Address
+
+	// NonCoopHint is set, if the sweep can not be spend cooperatively and
```
Fixed

starius commented on sweep/sweeper.go:
```diff
@@ -207,6 +207,30 @@ func (s *Sweeper) GetSweepFeeDetails(ctx context.Context,
 	// Calculate weight for this tx.
 	var weightEstimate input.TxWeightEstimator
+
+	// Add output.
```
You mean method GetSweepFeeDetails? It had name GetSweepFee in godoc. Fixed in a separate commit.

starius commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,308 @@
+		// Add the sweep to the batch virtually.
+		combinedBatch := batch.combine(sweep)
+
+		// The cost is fee increase.
```
Fixed

starius commented on sweepbatcher/greedy_batch_selection_test.go:
```diff
@@ -0,0 +1,684 @@
+			},
+			wantNewBatchFeeDetails: feeDetails{
+				FeeRate:       lowFeeRate,
+				CoopWeight:    lntypes.WeightUnit(445),
```
Decided to keep the fix of fee estimators in that PR.

Added consts. They simplified updating the test a lot! Thank you for suggestion! 🎉

hieblmi commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,308 @@
+// compares costs of adding the sweep to each existing batch, and costs of new
+// batch creation for this sweep and returns BatchId of the winning batch. If
+// the best option is to create a new batch, it returns specialId. Each fee
+// details has also IsExternalAddr flag. There is a rule that sweeps having flag
```
nvm, I think that wouldn't make sense since it is not guaranteed that the external address is part of our wallet, hence we can't sweep it.

hieblmi reviewed (approved):
LGTM 💾 Frequent loop-out users will love this feature!

GeorgeTsagk commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,315 @@
+// of each alternative. Based on the estimates of selectBatch(), this method
+// adds the sweep to the batch that results in the least overall fee increase,
+// or creates new batch for it. If this method fails for whatever reason, the
+// caller falls back to the simple algorithm (method handleSweep).
```
👍 good idea to keep the existing mechanism as fall back

GeorgeTsagk commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,315 @@
+// caller falls back to the simple algorithm (method handleSweep).
+func (b *Batcher) greedyAddSweep(ctx context.Context, sweep *sweep) error {
+	if b.customFeeRate == nil {
+		return errors.New("greedy batch selection algorithm requires " +
```
why is custom fee rate required? couldn't you ask the backend for a feerate and run the estimation based on that?

starius replied:
This was fixed in the following commits:

 - sweepbatcher: always fill sweep.minFeeRate
 - sweepbatcher: always try greedy batch selection

GeorgeTsagk commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -0,0 +1,315 @@
+
+	// Run the algorithm. Get batchId of the best batch, or newBatchSignal
+	// if the best option is to create new batch.
+	batchId, err := selectBatch(
+		batches, sweepFeeDetails, newBatchFeeDetails,
+	)
+	if err != nil {
+		return fmt.Errorf("batch selection algorithm failed for sweep "+
+			"%x: %w", sweep.swapHash[:6], err)
+	}
+
+	// If the best option is to start new batch, do it.
+	if batchId == newBatchSignal {
+		return b.spinUpNewBatch(ctx, sweep)
+	}
+
+	// Locate the batch to add the sweep to.
+	bestBatch, has := b.batches[batchId]
+	if !has {
+		return fmt.Errorf("batch selection algorithm returned "+
+			"batch id %d which doesn't exist, for sweep %x",
+			batchId, sweep.swapHash[:6])
+	}
+
+	// Add the sweep to the batch.
+	accepted, err := bestBatch.addSweep(ctx, sweep)
```
This may need to be enhanced

Currently, `addSweep` will also run checks that are related to the timeout window of the batch vs the sweep that is being added. See here https://github.com/lightninglabs/loop/blob/master/sweepbatcher/sweep_batch.go#L440-L447

This means that the single best solution that the greedy algorithm has calculated may be rejected because the timeout heights of the sweeps are too far away. I believe this is not the desired behavior?

I see 2 possible paths here:

a) Create a version of `addSweep` that does not check for the timeout window limits, allowing swaps with very distant timeout heights to be in the same batch
b) Enhance algorithm to return an array of "best batches", sorted from best to worst. Then on the above line (`76`) we would have to iterate over the results and try to `addSweep` in order from the best to worst option

hieblmi replied:
How often does it happen that a sweep is rejected based on the timeout? If that's not a common case then we can KISS and just try and add to the selected greedy batch, and if timeout prevents addition then fail the greedy algo and fall back to the default. Wdyt?

GeorgeTsagk replied:
> Didn't see your c), is a high failure rate expected?

That highly depends on the type of looper, frequent loopers will probably encounter this more often, as they are the ones having many batches running in parallel

starius replied:
I like the option b) Enhance algorithm to return an array of "best batches", sorted from best to worst.

The selection algorithm already has a metric by which is compares all the alternatives, so it is easy to extend to return a sorted array.

starius replied:
I implemented it and pushed a temporary commit with this change. Please take a look! I'll squash it before merging.

GeorgeTsagk commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -44,43 +48,48 @@ func (b *Batcher) greedyAddSweep(ctx context.Context, sweep *sweep) error {
+			return nil
+		}
+
+		log.Infof("Batch selection algorithm returned batch id %d for "+
```
nit: I think it's better to do `Debugf` here

starius replied:
Done :+1:

GeorgeTsagk commented on sweepbatcher/greedy_batch_selection.go:
```diff
@@ -294,16 +314,28 @@ func selectBatch(batches []feeDetails, sweep, oneSweepBatch feeDetails) (int32,
+	// Sort the alternatives by cost. The lower the cost, the better.
+	sort.Slice(alternatives, func(i, j int) bool {
+		return alternatives[i].cost < alternatives[j].cost
+	})
```
beautiful :100:

GeorgeTsagk reviewed (approved):
awesome :fire: 
LGTM
